### PR TITLE
Added create_open and close_add functions

### PR DESCRIPTION
**Pull request by Lnjacobs**
'create_open': creates a room where anyone (in the game or not) can add (but not remove rolls)

'close_add': takes an open room and closes it so only the moderator can add roles.. incase of trolling

added attribute 'rooms.open' which is assumed 'False' unless passed 'True'

added to 'reveal' function so it would switch the 'room.started' to 'False' added to the 'add' function to consider if a room was 

open before rejecting a non-mod from adding roles

### CODE REVIEW

**Comment by sarangkarpate** [comment]
**Meta note**:

Let's remove the `.idea` folder please - I see it's adding new files as well

File: role_assignment.py [hunk 1]
```diff
@@ -300,5 +377,21 @@ async def update_moderator(ctx, *args):
             {get_user(ctx, player).display_name: role for player, role in Rooms[room].assigned_roles.items()})
         await get_user(ctx, Rooms[room].moderator).send(game_description)
 
+@bot.command(name="close_add", brief="<room> makes so only moderator can add roles",
```
Comment (sarangkarpate) [inline role_assignment.py:380]: Can we change this to `restrict`? `close_add` seems a bit wordy

File: config.json [hunk 2]
```diff
@@ -1,4 +1,4 @@
 {
-  "Token": "<PlaceHolder>",
-  "Prefix": "!"
+  "Token": "<Place holder>",
+  "Prefix": "$"
```
Comment (sarangkarpate) [inline config.json:3]: Revert to "!" - production prefix should remain the same.

File: role_assignment.py [hunk 3]
```diff
@@ -48,8 +57,34 @@ async def create_room(ctx, *args):
...
+    button = discord.ui.Button(label="Join village", emoji="🐺", style=discord.ButtonStyle.green)
+    async def button_callback(interaction):
+        if interaction.user.id in Rooms[room].players:
+            await interaction.user.send("You are already in room %s" % room)
+            return
+        Rooms[room].add_player(interaction.user.id)
+        await interaction.user.send("You have joined room %s" % room)
+        await interaction.response.edit_message(content="Join the new %s village by clicking here! %s has joined"
+                                                        % (room, interaction.user.display_name))
+    button.callback = button_callback
```
Comment (sarangkarpate) [inline role_assignment.py:71]: **This can be super useful to others if abstracted**

-----

If we can abstract away the code to create a button in a new method like `create_button`, it'll be really useful & reusable. The method can take in inputs like:
1. Button label
2. Emoji
3. Button Callback function

and return a `view` that can be included in `ctx.send`. We can add this helper method in `utils`.

File: role_assignment.py [hunk 4]
```diff
@@ -129,6 +182,23 @@ async def invite_player(ctx, *args):
     for p in ctx.message.mentions:
         await p.send("You have been added to room: %s" % room)
 
+@bot.command(name="invite_offline", brief="<room> <player1> <player2> ... Adds person/people without dicrod toa  room",
```
Comment (sarangkarpate) [inline role_assignment.py:185]: Typos in `brief`

File: role_assignment.py [hunk 5]
```diff
@@ -129,6 +182,23 @@ async def invite_player(ctx, *args):
...
+             description="Use this command to include players without discord. There role will be sent to the moderator."
```
Comment (sarangkarpate) [inline role_assignment.py:186]: Typo: `There` to `Their`

File: role_assignment.py [hunk 6]
```diff
@@ -129,6 +182,23 @@ async def invite_player(ctx, *args):
...
+                         "Syntax: invite <room_name> <player1> <player2> ...")
+async def invite_player_offline(ctx, *args):
```
Comment (sarangkarpate) [inline role_assignment.py:188]: *Nit*: can be renamed to `invite_offline_player`

File: role_assignment.py [hunk 7]
```diff
@@ -129,6 +182,23 @@ async def invite_player(ctx, *args):
...
+    if not args:
+        await ctx.author.send("<room name> is a required argument to the !invite command.")
+        return
+    room = args[0]
+    if not room in Rooms:
```
Comment (sarangkarpate) [inline role_assignment.py:193]: *nit: change to `if room not in Rooms` for better readability.

File: role_assignment.py [hunk 8]
```diff
@@ -129,6 +182,23 @@ async def invite_player(ctx, *args):
...
+        await ctx.author.send("A room with the name '%s' doesn't exist." % room)
+        return
+    print(args)
```
Comment (sarangkarpate) [inline role_assignment.py:196]: Unnecessary `print`

File: role_assignment.py [hunk 9]
```diff
@@ -129,6 +182,23 @@ async def invite_player(ctx, *args):
     for p in ctx.message.mentions:
         await p.send("You have been added to room: %s" % room)
 
+@bot.command(name="invite_offline", brief="<room> <player1> <player2> ... Adds person/people without dicrod toa  room",
+             description="Use this command to include players without discord. There role will be sent to the moderator."
+                         "Syntax: invite <room_name> <player1> <player2> ...")
+async def invite_player_offline(ctx, *args):
+    if not args:
```
Comment (sarangkarpate) [inline role_assignment.py:189]: Given this code (line 189 - 196) is reused both in `invite` and `invite_player_offline` let's abstract it in a method and reuse it in both places. We can call it as `invite_validate`

File: role_assignment.py [hunk 10]
```diff
@@ -129,6 +182,23 @@ async def invite_player(ctx, *args):
...
+        await ctx.author.send("<room name> is a required argument to the !invite command.")
+        return
+    room = args[0]
+    if not room in Rooms:
+        await ctx.author.send("A room with the name '%s' doesn't exist." % room)
+        return
+    print(args)
+    Rooms[room].add_players([p for p in args[1:]])
```
Comment (sarangkarpate) [inline role_assignment.py:197]: Need checks for args[1:] existence - else it can lead to weird message in Line 199

File: server/backend.py [hunk 11]
```diff
@@ -18,12 +18,13 @@ def __hash__(self) -> int:
...
 
 class Room():
-    def __init__(self, moderator: int):
+    def __init__(self, moderator: int, open=False):
         self.roles = {}
         self.players = set()
         self.moderator = moderator
         self.started = False
         self.assigned_roles = {}
+        self.open = open
```
[the remark this answers is not in the document]
Comment (sarangkarpate) [reply server/backend.py:27]: Also, can we rename this to `self.restricted` - I got confused a bunch of times thinking it meant started.

File: role_assignment.py [hunk 12]
```diff
@@ -142,7 +212,7 @@ async def add_role(ctx, *args):
     if not room in Rooms:
         await ctx.author.send("A room with the name '%s' doesn't exist." % room)
         return
-    if not ctx.message.author.id == Rooms[room].moderator:
+    if not ctx.message.author.id == Rooms[room].moderator and Rooms[room].open == False:
```
Comment (sarangkarpate) [inline role_assignment.py:215]: If changing `.open` to `.restricted` - please change code check to:
```
not Rooms[room].restricted
```

File: role_assignment.py [hunk 13]
```diff
@@ -203,10 +274,10 @@ async def assign_roles(ctx, *args):
         return
     try:
         game = village.start_game()
-        game_description = pretty_print_dictionary({get_user(ctx, player).display_name: role for player, role in game.items()})
-        await get_user(ctx, village.moderator).send(game_description)
+        await get_user(ctx, village.moderator).send(pretty_print_dictionary(game_dict(ctx, game.items())))
```
Comment (sarangkarpate) [inline role_assignment.py:277]: Nice - me likey!

File: role_assignment.py [hunk 14]
```diff
@@ -203,10 +274,10 @@ async def assign_roles(ctx, *args):
...
         for player, role in game.items():
-            await get_user(ctx, player).send("Game: %s, Role: %s" % (room, role))
+            if isinstance(player, str) != True:
```
Comment (sarangkarpate) [inline role_assignment.py:279]: Curious: what is this check for?

Comment (sarangkarpate) [reply role_assignment.py:279]: To accomodate offline players?

File: role_assignment.py [hunk 15]
```diff
@@ -224,9 +295,15 @@ async def list_room_info(ctx, *args):
...
+    player_names = []
+    for player in Rooms[room].players:
+        try:
+            player_names.append(get_user(ctx, player).display_name)
+        except ValueError:
+            player_names.append(player)
     await ctx.send("Room: %s\nModerator: %s\nPlayers: %s\nRoles\n-----------\n%s" % (
-        room, get_user(ctx, int(Rooms[room].moderator)).display_name,
-        str([get_user(ctx, x).display_name for x in Rooms[room].players]), str(Rooms[room].get_role_information())))
+        room, get_user(ctx, int(Rooms[room].moderator)).display_name, player_names,
```
Comment (sarangkarpate) [inline role_assignment.py:305]: *Nit*: send `sorted(player_names)` instead of `player_names` for easier readability

File: role_assignment.py [hunk 16]
```diff
@@ -255,10 +332,9 @@ async def reveal_room_info(ctx, *args):
     if not ctx.message.author.id == Rooms[room].moderator:
         await ctx.author.send("You must be the moderator for the room to reveal roles.")
         return
-
-    game_mapping = {get_user(ctx, player).display_name: role for player, role in Rooms[room].assigned_roles.items()}
-    game_description = pretty_print_dictionary(game_mapping)
+    game_description = pretty_print_dictionary(game_dict(ctx,Rooms[room].assigned_roles.items()))
```
Comment (sarangkarpate) [inline role_assignment.py:335]: Me likey again.

File: role_assignment.py [hunk 17]
```diff
@@ -255,10 +332,9 @@ async def reveal_room_info(ctx, *args):
...
     await ctx.send(game_description)
```
Comment (sarangkarpate) [inline role_assignment.py:336]: Let's also add a message saying that the room is now closed / (not started) if possible.

File: role_assignment.py [hunk 18]
```diff
@@ -48,8 +57,34 @@ async def create_room(ctx, *args):
...
+        await interaction.response.edit_message(content="Join the new %s village by clicking here! %s has joined"
+                                                        % (room, interaction.user.display_name))
+    button.callback = button_callback
+    view = discord.ui.View()
+    view.add_item(button)
+    await ctx.send("Join the new %s village by clicking here!" % room, view=view)
+
     await get_moderator(ctx, room).send("Room %s created." % room)
 
+@bot.command(name="create_open", brief="<room> Creates a room (village) where anyone can add roles",
```
[the remark this answers is not in the document]
Comment (Lnjacobs) [reply role_assignment.py:78]: It depends on what the moderator wants. Create is still useful if someone wants to add the roles themselves. This just provides another option.

File: role_assignment.py [hunk 19]
```diff
@@ -107,6 +142,24 @@ async def remove_player(ctx, *args):
     for p in ctx.message.mentions:
         await p.send("You have been removed from room: %s" % room)
 
+@bot.command(name="remove_offline", brief="<room> <player1> <player2> ... Removes offline players from the room",
+             description="Use this command to remove players not on discord from a room."
+                         "Syntax: remove <room_name> <player1> <player2> ...")
```
[the remark this answers is not in the document]
Comment (Lnjacobs) [reply role_assignment.py:147]: I don't think offline means what you think it means. I'm talking about people who don't have discord or aren't in our server. So if someone shows up without a phone/dead phone/doesn't have discord we can still include them in the game. 

It has nothing to do with the online status of the discord users. I just thought offline was a nicer name than dummy (since we are inviting a dummy person to the game)

File: server/backend.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Lnjacobs) [reply server/backend.py:27]: I like open more for the command because its a lot less to type then restricted. I can rename the commands and attributes in the function to restricted tho.

File: role_assignment.py (hunk 14 above)
Comment (Lnjacobs) [reply role_assignment.py:279]: To accommodate players who don't have discord. Ie-if the players string and game dictionary, they are a string and not a discord user class. Thus we cannot message them with causing errors.

File: config.json [hunk 20]
```diff
@@ -1,4 +1,4 @@
 {
-  "Token": "<PlaceHolder>",
-  "Prefix": "!"
+  "Token": "<Place holder>",
```
[the remark this answers is not in the document]
Comment (Lnjacobs) [reply config.json:2]: I didnt realize it needed to be explicit. Will do now.

File: role_assignment.py (hunk 15 above)
Comment (Lnjacobs) [reply role_assignment.py:305]: Woops. I'll add it back. I also am planning to add a count for the !list command so we can quantify the number of roles and players without counting

File: role_assignment.py (hunk 17 above)
Comment (Lnjacobs) [reply role_assignment.py:336]: Do you just mean a message in the existing chat room? Or do you want to send this message to anyone trying to add roles/invite players? If we do it the second way, then we would have to add a recycle function (which I believe you didn't want to so im now assuming you meant in the general chat but am still looking for confirmation... woo run on sentences)

File: role_assignment.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (sarangkarpate) [reply role_assignment.py:78]: I feel like having both ‘create’ and ‘create_open’ can create some confusion initially. I’d much rather have a consistent experience where either a room is created with restricted roles / open roles and then have a separate command to open / restrict the room

File: role_assignment.py (hunk 14 above)
Comment (sarangkarpate) [reply role_assignment.py:279]: Can we rewrite this to 
‘not isInstance(player, str)’

and add a comment that str contents are only for non discord players

File: role_assignment.py (hunk 15 above)
Comment (sarangkarpate) [reply role_assignment.py:305]: Yes please, that would really help!

File: role_assignment.py (hunk 17 above)
Comment (sarangkarpate) [reply role_assignment.py:336]: Hehe, fine with either

File: server/backend.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (sarangkarpate) [reply server/backend.py:27]: We can keep the command as ‘open’ but we should just change the variable name to restricted because it helps understand the code a lot more. We can also keep the command as ‘restrict’ and not ‘restricted’ but I’m not too particular about that. (We can keep the command open)

[the remark this answers is not in the document]
Comment (sarangkarpate) [reply server/backend.py:27]: Lol so basically what you just said haha - sorry I’m reviewing your code and having a conversation with someone at the same time

[the remark this answers is not in the document]
Comment (sarangkarpate) [reply server/backend.py:27]: Lol so basically what you just said haha - sorry I’m reviewing your code and having a conversation with someone at the same time

File: role_assignment.py (hunk 14 above)
Comment (Lnjacobs) [reply role_assignment.py:279]: Will add

File: CreateButtons.py [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+        "prince": ["👑", discord.ButtonStyle.green],
+        "beholder": ["🤐", discord.ButtonStyle.green],
+        "masons": ["👩‍🚒👨‍🚒", discord.ButtonStyle.green],
+        "thing": ["👈👾👉", discord.ButtonStyle.green],
+        "chupacabra": ["👹", discord.ButtonStyle.green],
+        "tough guy": ["💪", discord.ButtonStyle.green],
+        "wheelsmith": ["🤸‍♀️", discord.ButtonStyle.green],
+        "doctor": ["🩺", discord.ButtonStyle.green],
+        "body guard": ["💂‍♀️", discord.ButtonStyle.green],
+        "witch": ["🎆", discord.ButtonStyle.green],
```
Comment (NickJosephson) [inline CreateButtons.py:98]: Potential witch emoji alternatives: 🧙🧙‍♀️🧪

File: role_assignment.py [hunk 22]
```diff
@@ -36,6 +36,15 @@ def get_moderator(ctx, room_name):
         # For DM scenarios
         return ctx.author
 
+def game_dict(ctx,items):
```
[the remark this answers is not in the document]
Comment (sarangkarpate) [reply role_assignment.py]: Yes utils.py under util folder

File: CreateButtons.py (hunk 21 above)
Comment (Lnjacobs) [reply CreateButtons.py:98]: everyone is a critic :(

Comment (Lnjacobs) [reply CreateButtons.py:98]: Also these are no properly implemented lol. Still working on it.

File: role_assignment.py (hunk 17 above)
Comment (Lnjacobs) [reply role_assignment.py:463]: I went with the first option, so no recycle function

File: role_assignment.py (hunk 6 above)
Comment (Lnjacobs) [reply role_assignment.py:266]: that seems wordy

File: role_assignment.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Lnjacobs) [reply role_assignment.py]: Okay we have create (defaults to closed), open, and restricted now

**Review by gingrichian — APPROVED** [review]

**Comment by NickJosephson** [comment]
Nice work coder!
